### PR TITLE
Scalability issues

### PR DESCRIPTION
**Pull request by paulgirard**
See the related PR in tableschema-py [https://github.com/frictionlessdata/tableschema-py/pull/254](https://github.com/frictionlessdata/tableschema-py/pull/254)

I have scalability issues when trying to validate a datapackage which contains one resource with 397201 lines containing foreign keys.
I needed to split my 397k lines resource into many files to organize those by source.
I finally chose to use share schema and group notion to split into many resources.
Trying to validate all those lines brought many scalability issues in both tableschema and datapackage libraries.

First issue is about memory management.
Checking relations in a resource, makes the object not only to read the related resource but to hold the data in memory as it is kept as an object attribute.
This has the consequences to make the memory grow when checking relations of a large amount of resource.
I don't know why the relations data are kept in the object so I proposed in this PR a new method to clean the relations data *drop_relations*.

Once the memory issue solved I had a performance one. Checking relations of my +1000 resources hodling 397k lines of data took 98m49.895s.
That's very long.

So I thought about two optimizations.
The first one is to avoid loading the relations for each resources which belongs to a group.
To enable this optimization I propose to add a *check_relations* method into the Group object.
Thus we load relations data once and then use this data in memory to validate all the resource belonging to that group.

Than a second optimization has been proposed into tableschema-py  by the PR tableschema-py [https://github.com/frictionlessdata/tableschema-py/pull/254](https://github.com/frictionlessdata/tableschema-py/pull/254). 
The idea is to pre-index the relations data by the values of the foreign keys. This index is called foreign_keys_values. This index is then used to test if the row reference some of the existing value (simple hash map lookup).

To speed up, I also propose to expose the *get_foreign_keys_values()* method so that the Group object can pre-compute the index only once before using it to validate all the resource.

Using those optimizations made the validation process to drop from 98m49.895s to 1m3.609s.

I've just realized that this PR miss updates on the documentation.
Let's see if the principles are good by reviewers before updating the doc..

@roll what do you think ?

### CODE REVIEW

**Comment by paulgirard** [comment]
Oh yeah of course since this PR is based on an update of the tableschema-py dependency, CI failes.
This PR can be valid before the tablechema one is accepted and deployed...

File: datapackage/group.py [hunk 1]
```diff
@@ -52,3 +53,24 @@ def read(self, limit=None, **options):
             if count == limit:
                 break
         return rows
+
+    def check_relations(self):
```
Comment (roll) [inline datapackage/group.py]: Could you please add a test for this method - https://coveralls.io/builds/26209815/source?filename=datapackage%2Fgroup.py#L62?

File: datapackage/group.py [hunk 2]
```diff
@@ -52,3 +53,24 @@ def read(self, limit=None, **options):
...
+
+        # alternative to check_relations from tableschema-py
+        for resource in self.__resources:
+            try:
+                resource.check_relations(foreign_keys_values=foreign_keys_values)
+            except exceptions.DataPackageException as exception:
+                print('in %s: ' % resource.name)
+                if exception.multiple:
+                    for error in exception.errors:
+                        print(error)
```
[the remark this answers is not in the document]
Comment (paulgirard) [reply datapackage/group.py]: Indeed, raising looks like the good thing to do.

File: datapackage/group.py [hunk 3]
```diff
@@ -52,3 +53,24 @@ def read(self, limit=None, **options):
...
+                else:
+                    print(exception)
```
[the remark this answers is not in the document]
Comment (paulgirard) [reply datapackage/group.py]: Indeed, raising looks like the good thing to do

File: datapackage/group.py (hunk 1 above)
Comment (paulgirard) [reply datapackage/group.py]: Yep almost done.

**Comment by paulgirard** [comment]
@roll I just removed my try catch to let exceptions flow upstream. Maybe there is a policy to raise a datapackage exceptions and not tableschema ones ?

**Pull request merged**